### PR TITLE
ECL-179: Added pages relating to AML regulated activity

### PR DESCRIPTION
**Pull request by TonyCOgilvie**

### CODE REVIEW

File: app/uk/gov/hmrc/economiccrimelevyregistration/models/Registration.scala [hunk 1]
```diff
@@ -19,14 +19,18 @@ package uk.gov.hmrc.economiccrimelevyregistration.models
...
   internalId: String,
   entityType: Option[EntityType],
   meetsRevenueThreshold: Option[Boolean],
   amlSupervisor: Option[AmlSupervisor],
   incorporatedEntityJourneyData: Option[IncorporatedEntityJourneyData],
   soleTraderEntityJourneyData: Option[SoleTraderEntityJourneyData],
-  partnershipEntityJourneyData: Option[PartnershipEntityJourneyData]
+  partnershipEntityJourneyData: Option[PartnershipEntityJourneyData],
+  startedAmlRegulatedActivity: Option[Boolean],
+  amlStartDate: Option[LocalDate]
```
Comment (peter-hazell) [inline app/uk/gov/hmrc/economiccrimelevyregistration/models/Registration.scala:33]:
```suggestion
  startedAmlRegulatedActivityInCurrentFy: Option[Boolean],
  amlRegulatedActivityStartDate: Option[LocalDate]
```

File: app/uk/gov/hmrc/economiccrimelevyregistration/controllers/AmlStartDateController.scala [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import uk.gov.hmrc.economiccrimelevyregistration.navigation.AmlStartDatePageNavigator
+import uk.gov.hmrc.economiccrimelevyregistration.views.html.AmlStartDateView
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendBaseController
+
+import java.time.LocalDate
+import javax.inject.{Inject, Singleton}
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class AmlStartDateController @Inject() (
```
Comment (peter-hazell) [inline app/uk/gov/hmrc/economiccrimelevyregistration/controllers/AmlStartDateController.scala:36]: Think maybe we can name this a bit more explicitly e.g.

`AmlRegulatedActivityStartDateController`

File: app/uk/gov/hmrc/economiccrimelevyregistration/forms/AmlStartDateFormProvider.scala [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+class AmlStartDateFormProvider extends Mappings {
+  val dateIsWithinCurrentFinancialYear: Constraint[LocalDate] = Constraint[LocalDate] { date: LocalDate =>
+    if (
+      date.isBefore(EclTaxYear.currentFinancialYearStartDate) || date.isAfter(EclTaxYear.currentFinancialYearEndDate)
+    ) {
+      Invalid("amlStartDate.error.notWithinFinancialYear")
+    } else {
+      Valid
+    }
+  }
```
Comment (peter-hazell) [inline app/uk/gov/hmrc/economiccrimelevyregistration/forms/AmlStartDateFormProvider.scala:35]: I've realised we didn't actually need to write this ourselves

There is a constraint for min and max date already in our `Constraints` trait under the `mappings` package

Instead we can just do the following:

```  def apply(): Form[LocalDate] = Form(
    (
      "value",
      localDate(
        "amlStartDate.error.invalid",
        "amlStartDate.error.allRequired",
        "amlStartDate.error.twoRequired",
        "amlStartDate.error.required"
      ).verifying(
        minDate(EclTaxYear.currentFinancialYearStartDate, "amlStartDate.error.notWithinFinancialYear"),
        maxDate(EclTaxYear.currentFinancialYearEndDate, "amlStartDate.error.notWithinFinancialYear")
      )
    )
  )
```

File: test/uk/gov/hmrc/economiccrimelevyregistration/forms/AmlStartDateFormProviderSpec.scala [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    val fieldName   = "value"
+    val requiredKey = "amlStartDate.error.allRequired"
+
+    behave like dateField(
+      form,
+      fieldName,
+      datesBetween(EclTaxYear.currentFinancialYearStartDate, EclTaxYear.currentFinancialYearEndDate)
+    )
+
+    behave like mandatoryDateField(
```
Comment (peter-hazell) [inline test/uk/gov/hmrc/economiccrimelevyregistration/forms/AmlStartDateFormProviderSpec.scala:36]: Think we discussed in `mandatoryDateField` we could add some extra tests for different combinations of missing data

File: app/uk/gov/hmrc/economiccrimelevyregistration/models/Registration.scala (hunk 1 above)
Comment (TonyCOgilvie) [reply app/uk/gov/hmrc/economiccrimelevyregistration/models/Registration.scala:33]: 👍

File: app/uk/gov/hmrc/economiccrimelevyregistration/controllers/AmlStartDateController.scala (hunk 2 above)
Comment (TonyCOgilvie) [reply app/uk/gov/hmrc/economiccrimelevyregistration/controllers/AmlStartDateController.scala:36]: 👍

File: app/uk/gov/hmrc/economiccrimelevyregistration/forms/AmlStartDateFormProvider.scala (hunk 3 above)
Comment (TonyCOgilvie) [reply app/uk/gov/hmrc/economiccrimelevyregistration/forms/AmlStartDateFormProvider.scala:35]: 👍

File: test/uk/gov/hmrc/economiccrimelevyregistration/forms/AmlStartDateFormProviderSpec.scala (hunk 4 above)
Comment (TonyCOgilvie) [reply test/uk/gov/hmrc/economiccrimelevyregistration/forms/AmlStartDateFormProviderSpec.scala:36]: 👍

File: app/uk/gov/hmrc/economiccrimelevyregistration/navigation/AmlRegulatedPageNavigator.scala [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import play.api.mvc.Call
+import uk.gov.hmrc.economiccrimelevyregistration.controllers.routes
+import uk.gov.hmrc.economiccrimelevyregistration.models.Registration
+
+class AmlRegulatedPageNavigator extends PageNavigator {
+
+  override protected def navigateInNormalMode(registration: Registration): Call =
+    registration.startedAmlRegulatedActivityInCurrentFy match {
+      case Some(true)  => routes.AmlRegulatedActivityStartDateController.onPageLoad()
+      case Some(false) => ???
```
[the remark this answers is not in the document]
Comment (TonyCOgilvie) [reply app/uk/gov/hmrc/economiccrimelevyregistration/navigation/AmlRegulatedPageNavigator.scala]: 👍

File: test/uk/gov/hmrc/economiccrimelevyregistration/forms/behaviours/DateBehaviours.scala [hunk 6]
```diff
@@ -41,6 +47,32 @@ class DateBehaviours extends FieldBehaviours {
...
+          s"$key.month" -> "MM",
+          s"$key.year"  -> "yyyy"
+        )
+
+        val result = form.bind(data)
+
+        result.errors should contain only FormError(key, invalidKey, errorArgs)
+      }
+
+      "fail to bind data out-of-range" in {
```
Comment (peter-hazell) [inline test/uk/gov/hmrc/economiccrimelevyregistration/forms/behaviours/DateBehaviours.scala:64]: I think I understand this test - but I think the description is a little misleading, it's not actually testing anything to do with the date range (min/max date). It's testing that if you provide an invalid day, month, or year then it shouldn't bind even if they are valid 'integers'.

the test is very similar to the one above which is testing that it shouldn't bind if the day/month/year is not even an integer, i think you could combine them if you defined an `arbInvalidLocalDate` which generated 3 strings, the day, month and year

Comment (TonyCOgilvie) [reply test/uk/gov/hmrc/economiccrimelevyregistration/forms/behaviours/DateBehaviours.scala]: 👍

**Pull request merged**